lib: FRR next-hop yang data model.

Spantik opened the pull request:
A common nexthop and group nexthop yang data model
for all protocols in FRR.

Signed-off-by: Santosh P K <sapk@vmware.com>
Signed-off-by: Vishaldhingra <vdhingra@vmware.com>

Spantik commented:
Modified common next-hop and next-hop group for other protocols and zebra to use.

For next-hop would look like below with an example if used as leafref
Route -> 1.1.1.1 with nexthop 10.10.10.10, 20.20.20.20, 30.30.30.30
Route -> 2.2.2.2 with nexthop 20.20.20.20, 30.30.30.30, 40.40.40.40

Nexthop table:
id              Entry
--              ------
1                10.10.10.10
2                20.20.20.20
3                30.30.30.30
4                40.40.40.40

Nexthop Group table:
id                Entry
--                -----
10                1, 2, 3
20                2,3 ,4


Route with Nexthop will look like 
1.1.1.1/32 nexthop 1,2,3
2.2.2.2/32 nexthop 2,3,4

Route with Nexthop group would look like

1.1.1.1/32 nexthop 10
2.2.2.2/32 nexthop 20

Nexthop tree:
--------------
<img width="643" alt="Screenshot 2019-12-17 at 8 00 54 PM" src="https://user-images.githubusercontent.com/50885001/71004700-3f14bc80-2108-11ea-85f6-629e37a48416.png">

mjstapp commented on yang/frr-nexthop.yang:
```diff
@@ -100,96 +105,95 @@ module frr-nexthop {
-      type empty;
+    leaf active {
+      type boolean;
+      default true;
```
why is the default `true` here - why is this attribute different from others?

Spantik replied:
all the nexthops added are active by default? What would you prefer as default value here?

mjstapp commented on yang/frr-nexthop.yang:
```diff
@@ -100,96 +105,95 @@ module frr-nexthop {
-        "The nexthop-group name.";
-    }
-
+      "FRR next-hop.";
```
"next-hop" - is that intended - shouldn't it be "nexthop" ?

Spantik replied:
Done.

mjstapp commented on yang/frr-nexthop.yang:
```diff
@@ -200,10 +204,45 @@ module frr-nexthop {
+        leaf group-sub-id {
+          type frr-nexthop-ref;
+          description
+            "A refrence to a staticd-nexthop entry";
```
"refrence" - typo?

"staticd" - also a typo?

Spantik replied:
Will correct it.

mjstapp commented on yang/frr-nexthop.yang:
```diff
@@ -200,10 +204,45 @@ module frr-nexthop {
+        type uint32;
+        description
+          "Identifies a nexthop within a nexthop group; the entries
+           are ordered by id value, and the value has no other meaning.";
```
The description text here doesn't seem right - it seems to be talking about a single nexthop, rather than a group. Also, the id value in a 'real' group does have some meaning - at least, it can be used when interacting with the newer linux kernels.

Spantik replied:
I will update the same.

mjstapp commented on yang/frr-nexthop.yang:
```diff
@@ -200,10 +204,45 @@ module frr-nexthop {
+      }
+      list group-nexthop {
+        key "group-sub-id";
+        leaf group-sub-id {
```
Don't we need to have some kind of index key concept here, to manage the embedded list of nexthops?

Spantik replied:
this is a id manging set of id's. if you do not have a group nexthop then all you have is a set of nexthops ids alone like I explained in exmaple.

mjstapp commented on yang/frr-nexthop.yang:
```diff
@@ -200,10 +204,45 @@ module frr-nexthop {
           "Identifies a nexthop within a nexthop group; the entries
            are ordered by id value, and the value has no other meaning.";
```
again, the id value of a nexthop does have meaning now, in the relationships among individual nexthops and groups zebra composes them currently.

Spantik commented:
> fix the signoff line, how to do this is very well documented in the linux style docs, the linux git log, the frr git log, git even has an option for it `-s` - it is getting tiresome spending my time checking this on all these commits. I do not want to keep nacking prs for this.

@qlyoung we checked linux documenation and we are taking care of the same for all PRs. Sorry about this.

Spantik commented:
I have raised a new PR https://github.com/FRRouting/frr/pull/5578 after addressing most of the comments and had to abonden it due to rebase issue I faced.
